include/cocaine/framework/receiver.hpp

```diff
@@ -0,0 +1,324 @@
+#pragma once
+
+#include <functional>
+#include <string>
+#include <system_error>
+#include <queue>
+
+#include <boost/mpl/at.hpp>
+#include <boost/mpl/lambda.hpp>
+#include <boost/mpl/for_each.hpp>
+#include <boost/mpl/front.hpp>
+#include <boost/mpl/size.hpp>
+#include <boost/mpl/transform.hpp>
+#include <boost/mpl/vector.hpp>
+#include <boost/variant.hpp>
+
+#include <cocaine/common.hpp>
+#include <cocaine/idl/primitive.hpp>
+#include <cocaine/idl/streaming.hpp>
+#include <cocaine/rpc/asio/decoder.hpp>
+#include <cocaine/tuple.hpp>
+
+#include "cocaine/framework/error.hpp"
+#include "cocaine/framework/forwards.hpp"
+#include "cocaine/framework/message.hpp"
+
+namespace cocaine {
+
+namespace framework {
+
+namespace detail {
+
+/*!
+ * Transforms a typelist sequence into a single movable argument type.
+ *
+ * If the sequence contains a single element of type T, than the result type will be T.
+ * Otherwise the sequence will be transformed into a tuple.
+ *
+ * sequence<T, U...> -> std::tuple<T, U...>,
+ * sequence<T>       -> T.
+ */
+template<class U, size_t = boost::mpl::size<U>::value>
+struct packable {
+    typedef typename tuple::fold<U>::type type;
+};
+
+template<class U>
+struct packable<U, 1> {
+    typedef typename boost::mpl::front<U>::type type;
+};
+
+/*!
+ * primitive<T> -> variant<value<T>::type, error::type>
+ *              -> variant<T, tuple<int, string>>
+ * streaming<T> -> variant<chunk<T>::type, error::type, choke::type>
+ *              -> variant<T, tuple<int, string>, tuple<>>
+ */
```

Comment:
Hmm, it should produce a variant with duplicate types, which should compile, I suppose, but how to work with it - is the question.

We can wrap it into a strong typedef to avoid ambiguity.